### PR TITLE
[CELEBORN-1490][CIP-6] Extends FileMeta to support hybrid shuffle

### PR DESCRIPTION
**Pull request by reswqa**
### What changes were proposed in this pull request?

Second PR to support Hybrid Shuffle.

This extends `FileMeta` as hybrid shuffle supports reading while writing at segment-grained. 

We also introduce `isSegmentGranularityVisible` to some message and method to identify segment-based shuffle.

### Does this PR introduce _any_ user-facing change?
no.

### How was this patch tested?
no need.

### CODE REVIEW

File: common/src/main/java/org/apache/celeborn/common/meta/MapFileMeta.java [hunk 1]
```diff
@@ -17,16 +17,38 @@
...
+import java.util.List;
+import java.util.Map;
+
 public class MapFileMeta implements FileMeta {
   private int bufferSize;
   private int numSubPartitions;
   private String mountPoint;
+  private Map<Integer, Integer> subPartitionWritingSegmentId;
+  private List<SegmentIndex> subPartitionSegmentIndexes;
+  private volatile boolean hasWriteFinished;
```
Comment (RexXiong) [inline common/src/main/java/org/apache/celeborn/common/meta/MapFileMeta.java:32]: hasWriteFinished -> isWriterClosed? Seems this field also need persistent for recovery.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java [hunk 2]
```diff
@@ -164,17 +165,29 @@ public void setupLifecycleManagerRef(RpcEndpointRef endpointRef) {
...
+          "Shuffle data is empty for shuffle {} partitionId {} isSegmentGranularityVisible {}.",
+          shuffleId,
+          partitionId,
+          isSegmentGranularityVisible);
+      if (isSegmentGranularityVisible) {
+        // When the downstream reduce tasks start early than upstream map tasks, the shuffle
+        // partition locations may be found empty, should retry until the upstream task started
+        return CelebornBufferStream.empty();
+      } else {
+        throw new PartitionUnRetryAbleException(partitionId + " may be lost.");
```
Comment (SteNicholas) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java:188]:
```suggestion
        throw new PartitionUnRetryAbleException(String.format("Shuffle data lost for shuffle %d partition %d.", shuffleId, partitionId));
```

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java [hunk 3]
```diff
@@ -193,8 +206,30 @@ public CelebornBufferStream readBufferedPartition(
...
+   * Update the reduce file groups and obtain the PartitionLocations of the target
+   * shuffleId#partitionId. It is possible to return an empty array if the corresponding reduce file
+   * groups are nonexistent, a scenario likely arising when downstream reduce tasks are start early
+   * than upstream map tasks, e.g. Flink Hybrid Shuffle.
+   */
+  public PartitionLocation[] updateFileGroupAndGetLocations(
+      int shuffleId, int partitionId, boolean isSegmentGranularityVisible) throws IOException {
+    ReduceFileGroups fileGroups =
+        updateFileGroup(shuffleId, partitionId, isSegmentGranularityVisible);
+    if (fileGroups.partitionGroups == null
```
Comment (SteNicholas) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java:219]: Replace `fileGroups.partitionGroups == null || fileGroups.partitionGroups.isEmpty()` with `CollectionUtils.isEmpty(fileGroups.partitionGroups)`.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java [hunk 4]
```diff
@@ -193,8 +206,30 @@ public CelebornBufferStream readBufferedPartition(
...
+        || fileGroups.partitionGroups.isEmpty()
+        || !fileGroups.partitionGroups.containsKey(partitionId)) {
+      return new PartitionLocation[0];
+    } else {
+      PartitionLocation[] partitionLocations =
+          fileGroups.partitionGroups.get(partitionId).toArray(new PartitionLocation[0]);
+      return partitionLocations;
```
Comment (SteNicholas) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java:226]:
```suggestion
      return fileGroups.partitionGroups.get(partitionId).toArray(new PartitionLocation[0]);
```

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java [hunk 5]
```diff
@@ -516,6 +552,44 @@ public void regionFinish(int shuffleId, int mapId, int attemptId, PartitionLocat
...
+      int segmentId,
+      PartitionLocation location)
+      throws IOException {
+    final String mapKey = Utils.makeMapKey(shuffleId, mapId, attemptId);
+    final PushState pushState = pushStates.computeIfAbsent(mapKey, (s) -> new PushState(conf));
+    retrySendMessage(
+        () -> {
+          final String shuffleKey = Utils.makeShuffleKey(appUniqueId, shuffleId);
+          logger.debug(
+              "SegmentStart for shuffle {} map {} attemptId {} locationId {}.",
```
Comment (SteNicholas) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java:569]: Add `subPartitionId` and `segmentId` in log.

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 6]
```diff
@@ -786,8 +787,12 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
...
   private def handleGetReducerFileGroup(
       context: RpcCallContext,
-      shuffleId: Int): Unit = {
-    if (!registeredShuffle.contains(shuffleId)) {
+      shuffleId: Int,
+      isSegmentGranularityVisible: Boolean): Unit = {
+    // If isSegmentGranularityVisible is set to true, the downstream reduce task may start early than upstream map task, e.g. flink hybrid shuffle.
+    // Under these circumstances, there's a possibility that the shuffle might not yet be registered when the downstream reduce task send GetReduceFileGroup request,
+    // so we shouldn't send a SHUFFLE_NOT_REGISTERED response directly, should enqueue this request to pending list, and response to the downstream reduce task the ReduceFileGroup when the upstream map task register shuffle done
+    if (!registeredShuffle.contains(shuffleId) && !isSegmentGranularityVisible) {
```
Comment (SteNicholas) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:795]: When is `isSegmentGranularityVisible ` set to true?

File: common/src/main/java/org/apache/celeborn/common/meta/MapFileMeta.java [hunk 7]
```diff
@@ -49,11 +71,108 @@ public void setMountPoint(String mountPoint) {
...
+        subPartitionSegmentIndexes.add(new SegmentIndex(new HashMap<>()));
+      }
+    }
+    subPartitionSegmentIndexes
+        .get(subPartitionId)
+        .addFirstBufferIndexAndSegmentId(firstBufferIndex, segmentId);
+  }
+
+  public synchronized Integer getSegmentIdByFirstBufferIndex(int subPartitionId, int bufferIndex) {
+    if (subPartitionSegmentIndexes == null) {
```
Comment (SteNicholas) [inline common/src/main/java/org/apache/celeborn/common/meta/MapFileMeta.java:143]:
```suggestion
    if (subPartitionSegmentIndexes == null || subPartitionSegmentIndexes.size() < subPartitionId) {
```

Comment (SteNicholas) [reply common/src/main/java/org/apache/celeborn/common/meta/MapFileMeta.java:143]: Meanwhile, the return value could be simplified.

File: common/src/main/java/org/apache/celeborn/common/meta/MapFileMeta.java [hunk 8]
```diff
@@ -17,16 +17,38 @@
...
+import java.util.List;
+import java.util.Map;
+
 public class MapFileMeta implements FileMeta {
   private int bufferSize;
   private int numSubPartitions;
   private String mountPoint;
+  private Map<Integer, Integer> subPartitionWritingSegmentId;
+  private List<SegmentIndex> subPartitionSegmentIndexes;
+  private volatile boolean isWriterClosed;
```
Comment (SteNicholas) [inline common/src/main/java/org/apache/celeborn/common/meta/MapFileMeta.java:32]: `writer always closed as this is committed file info.`, when is the `isWriterClosed` false?

Comment (SteNicholas) [inline common/src/main/java/org/apache/celeborn/common/meta/MapFileMeta.java:32]: Where is `isWriteClosed` used?

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Review by RexXiong — APPROVED** [review]

**Pull request closed without merging**